### DISCUSSION TITLE
how do I render a midi file and load the audio data to memory?

### DISCUSSION THREAD

**Discussion by ghost** [FluidSynth/fluidsynth#922] [Q&A] [answered]
I'm using python. and what I've been doing lately is: using the CLI render the midi file to file x and then read the file and load it to memory. I know that the writing-reading routine is quite slow and in my use case efficiency is crucial. So I want to skip that part and just load the output to memory. How do I do that?

**Comment by mawe42** [comment, marked as answer]
Hi!

Assuming you are on Linux or a similar platform: If you specify `-` as filename, then FluidSynth will render the audio to stdout instead. Combine that with `-q` to make sure no extra output messes up your audio stream:
```
fluidsynth /path/to/font.sf2 /path/to/file.mid -q -F-
```

Then just execute that using Python and capture stdout, for example with `subprocess.check_output`.

Cheers

**Reply by ghost** [reply]
I'm in an ubuntu notebook, and there is no "q" option.

**Reply by ghost** [reply]
hold on, my fluidsynth version is 1.1.9. wtf?

**Reply by ghost** [reply]
I see, ubuntu 18.04 has only been updated to fluidsynth version 1.1.9. but can't I use the ubuntu 21 binaries or something?

**Reply by mawe42** [reply]
You could try to install the fluidsynth package from a newer Ubuntu release. Might work, might not. Or compile the latest fluidsynth yourself: https://github.com/FluidSynth/fluidsynth/wiki/BuildingWithCMake#building-on-linux

**Reply by ghost** [reply]
yeah, I did that. however, the returned stdout is empty. even without the "-q" flag. the command works perfectly fine if I specify a file name though.

Here is what I'm doing:
`stdout = subprocess.run(f"fluidsynth -f config.txt -r 22050 -T wav -F- {path}.mid".split(' '), stdout=PIPE).stdout`

running the command in the terminal yields this:
`RIF���WAVEfmt �����"V���X���data�����RIF���WAVEfmt �����"V���X���data����fluidsynth: error: Failed to open audio file '-' for writing
`

**Reply by ghost** [reply]
never mind, figured it out. Thanks for helping me all the same. for future reference see [this issue](https://github.com/FluidSynth/fluidsynth/issues/553)

**Reply by ghost** [reply]
hmmm, there seems to be no performance improvement. I suppose fluidsynth is still wring to the file stdout which I'm later reading from. Is there any way to speedup the synthesizing process? what if I used the api or something? does it have a function that returns the audio samples?

**Reply by mawe42** [reply]
We have an extensive API documentation, there is even an example on how to fast-render a MIDI file: https://www.fluidsynth.org/api/FileRenderer.html

You could also try to use `-o synth.dynamic-sample-loading=1` in the command-line. That prevents fluidsynth from loading the whole soundfont into memory and only load samples that are actually needed. That will probably speed up the rendering quite a bit. Otherwise I wouldn't expect a large performance increase by fast-rendering via API instead of the CLI.

**Reply by ghost** [reply]
I did see the docs. I looked for any function that returns the audio samples from a midi file and found none. Can I open an issue requesting it to be added?

**Reply by mawe42** [reply]
Not sure what you mean... MIDI files don't contain audio samples. If you mean a way to render a MIDI file to audio via the API (which is what you are currently doing via the CLI), then the above referenced example code is what you want.

**Reply by ghost** [reply]
yeah, I know. but like I said, the reading-writing routine is quite slow. I would like for a function that takes a given midi file, from it generate audio samples with the help of some soundfonts loaded with some settings and NOT write them to a file but return them as an array (or two in the case of stereo audio).

**Reply by mawe42** [reply]
If your main performance problem really is writing the synthesized audio to a file/pipe and reading the data into Python, then I suggest to ditch Python and go straight to C and use the audio rendering API to retrieve the samples: https://www.fluidsynth.org/api/group__audio__rendering.html

However... my guess is that you haven't done any profiling to find out where your performance bottleneck is. I would be *extremely* surprised if file I/O is your problem here.

**Reply by ghost** [reply]
tbh, I know that the file I/O is _slower_  than just loading to memory, I just don't know _how_ slow. but I need it to be as fast as possible so I can spend a few hours creating python bindings for some C functions (A language which I never used, but ditching python isn't an option). and for the record merely loading takes about .2 seconds while the rendering takes like 1.6 seconds so I'm hoping to get at least 12.5% faster rendering.

Anyways, thank you so much for helping me out. I really appreciate it. I hope you have a lovely day. :D
